### PR TITLE
Higher-level libobject API for objects with fixed scopes

### PR DESCRIPTION
**Pull request by maximedenes**
This is the very first step of a libobject cleanup. It should not change semantics.

This is still work in progress.

The next steps will be:
- A similar API for objects with locality
- Studying the few non-standard objects remaining
- Writing or fixing discharge functions for many objects
- Starting to unify the behaviors (either by making the meaning of `Local` and `Global` uniform, or by providing new attributes for scoping
- Removing "superglobals" (effects on `Require`)

This is essentially a libobject change, so the natural assignee is @gares (since @silene seems offline).

### CODE REVIEW

File: library/libobject.mli [hunk 1]
```diff
@@ -113,6 +113,51 @@ val classify_object : obj -> obj substitutivity
...
+We recommend to avoid declaring superglobal objects and using the nodischarge
+variants.
+*)
+
+val local_object : string ->
+  cache:(Libnames.object_name * 'a -> unit) ->
+  discharge:(Libnames.object_name * 'a -> 'a option) ->
+  'a object_declaration
+
+val local_object_nodischarge : string ->
```
Comment (SkySkimmer) [inline library/libobject.mli]: Why not make the discharge argument optional?

**Comment by maximedenes** [comment]
What this PR does it basically the first step towards making obvious how to answer https://github.com/coq/coq/wiki/Default-Locality-Behavior-of-Commands (in the code)

**Comment by ejgallego** [comment]
Let me very quickly suggest that you implement a `Print Objects` command so we can see the status of libobject at any time; this should be a debug command and could go away at some point.

**Comment by maximedenes** [comment]
> Let me very quickly suggest that you implement a Print Objects command so we can see the status of libobject at any time; this should be a debug command and could go away at some point.

It would dump all the registered libobjects and their kind (local, global, superglobal or ad-hoc), right?

**Comment by ejgallego** [comment]
> It would dump all the registered libobjects and their kind (local, global, superglobal or ad-hoc), right?

In fact the more it can dump the more we know the inversion of control is going away!

I'd like in fact to have this to understand a bit better how objects are sent to workers, etc...

File: library/libobject.mli [hunk 2]
```diff
@@ -113,6 +113,51 @@ val classify_object : obj -> obj substitutivity
...
+  subst:(Mod_subst.substitution * 'a -> 'a) option ->
+  'a object_declaration
+
+val superglobal_object : string ->
+  cache:(Libnames.object_name * 'a -> unit) ->
+  subst:(Mod_subst.substitution * 'a -> 'a) option ->
+  discharge:(Libnames.object_name * 'a -> 'a option) ->
+  'a object_declaration
+
+val superglobal_object_nodischarge : string ->
```
Comment (gares) [inline library/libobject.mli]: How can something be superglobal but not survive sections?

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I like the approach, the code change in most cases is trivial hence I think the list of owners is a bit too long.
Only the owners of libobject should be concerned.

File: library/libobject.mli (hunk 2 above)
Comment (maximedenes) [reply library/libobject.mli]: Do you mean "how" or "why"?

**Review by ppedrot — APPROVED** [review]
Dunno whether I'm supposed to be an owner, but I do like the approach.

**Comment by maximedenes** [comment]
> This is essentially a libobject change, so the natural assignee is @gares (since @silene seems offline).

@gares ping

**Comment by gares** [comment]
I'll be back on wednesday

**Comment by maximedenes** [comment]
> I'll be back on wednesday

@gares ping

**Comment by gares** [comment]
There is still a needs progress label, and I also think you should rebase.

**Comment by gares** [comment]
ping @maximedenes

**Comment by maximedenes** [comment]
@gares rebased

**Review by ejgallego — COMMENTED** [review]
This seems like a core API that has changed, however no entry in `dev/doc` ?

**Comment by maximedenes** [comment]
> This seems like a core API that has changed, however no entry in dev/doc ?

Right, I forgot to do it. It'll have to wait, though.

File: interp/declare.ml [hunk 3]
```diff
@@ -474,11 +474,10 @@ let assumption_message id =
...
 
 let input_universe_context : Univ.ContextSet.t -> Libobject.obj =
-  declare_object
-    { (default_object "Monomorphic section universes") with
-      cache_function = (fun (na, uctx) -> Global.push_context_set false uctx);
-      discharge_function = (fun (_, x) -> Some x);
-      classify_function = (fun a -> Dispose) }
+  declare_object @@ local_object "Monomorphic section universes"
+    ~cache:(fun (na, uctx) -> Global.push_context_set false uctx)
+    ~subst:None
```
Comment (SkySkimmer) [inline interp/declare.ml]: No `subst` argument to local_object ^^

Comment (maximedenes) [reply interp/declare.ml]: Ok, I'm even too tired to get a commit right, so I should clearly not be doing any technical work.

**Comment by amahboubi** [comment]
HI @maximedenes, my review is requested here but I do not feel competent enough on this code. I like the announced objective, but cannot make any useful comment at this stage.

**Comment by gares** [comment]
ping @maximedenes the doc that is missing does not seem to be a daunting task, your should be able to reuse the .mli comment.

**Comment by gares** [comment]
ping @maximedenes

**Comment by maximedenes** [comment]
> ping @maximedenes

Again, this is not on the top of my list at the moment, it will have to wait as written above.

**Comment by maximedenes** [comment]
> Well, the last message from you was on the 29th of August ;-)

Sure, no problem in pinging. I simply meant that it was still on my radar but we are very active on topics with higher priority.

**Comment by maximedenes** [comment]
Rebased and added documentation. Should be ready modulo CI.

**Pull request merged**